### ISSUE TITLE
-Q max_qps limit does not work as advertised

### ISSUE THREAD

**Issue by pspacek** [DNS-OARC/dnsperf#224]
Version affected: 2.10.0

First, definition of -Q in the manual page:
> -Q max_qps
> Limits the number of requests **per second**.  There is no default limit.

(Emphasis added by me.)

This does not match the real behavior of that feature. Right now the code looks like this:
```c
        /* Rate limiting */
        if (tinfo->max_qps > 0) {
            run_time = now - times->start_time;
            req_time = (MILLION * stats->num_sent) / tinfo->max_qps;
            if (req_time > run_time) {
                usleep(req_time - run_time);
                now = perf_get_time();
                continue;
            }
        }
```

It calculates when the next packet should be sent based on **total** number of packets, not number of packets in given second.

Consequently, if the server is slower than specified -Q for a period of time, unsent packets "accumulate" and then the -Q value is overshot, possibly very significantly.

Here is example of such test run in table. Columns 2, 3, 4 are number of packets as reported by dnsperf. See time 25 s and following.

| time [s] | sent   | recvd  | timeouts |
|----------|--------|--------|----------|
| 1        | 100021 | 99673  | 0        |
| 2        | 100031 | 100024 | 0        |
| 3        | 100028 | 100027 | 0        |
| 4        | 100027 | 100031 | 0        |
| 5        | 100028 | 100030 | 0        |
| 6        | 100023 | 100023 | 323      |
| 7        | 100022 | 100023 | 0        |
| 8        | 100018 | 100015 | 0        |
| 9        | 100025 | 100020 | 0        |
| 10       | 100019 | 100022 | 0        |
| 11       | 100130 | 100134 | 0        |
| 12       | 100026 | 100026 | 0        |
| 13       | 100028 | 100033 | 0        |
| 14       | 100018 | 100018 | 0        |
| 15       | 100032 | 100025 | 0        |
| 16       | 100023 | 100027 | 0        |
| 17       | 100026 | 100017 | 0        |
| 18       | 100022 | 100028 | 0        |
| 19       | 100025 | 100020 | 0        |
| 20       | 100026 | 100025 | 0        |
| 21       | 100034 | 100039 | 0        |
| 22       | 100026 | 100029 | 0        |
| 23       | 100028 | 100014 | 0        |
| 24       | 100023 | 100040 | 0        |
| 25       | 100028 | 100023 | 0        |
| 26       | 51631  | 50659  | 0        |
| 27       | 0      | 0      | 0        |
| 28       | 0      | 0      | 0        |
| 29       | 0      | 0      | 0        |
| 30       | 0      | 0      | 0        |
| 31       | 9383   | 8383   | 1000     |
| 32       | 0      | 0      | 0        |
| 33       | 0      | 0      | 0        |
| 34       | 0      | 0      | 0        |
| 35       | 0      | 0      | 0        |
| 36       | 9388   | 8388   | 1000     |
| 37       | 0      | 0      | 0        |
| 38       | 0      | 0      | 0        |
| 39       | 0      | 0      | 0        |
| 40       | 0      | 0      | 0        |
| 41       | 9349   | 8350   | 1000     |
| 42       | 9      | 8      | 0        |
| 43       | 0      | 0      | 0        |
| 44       | 0      | 0      | 0        |
| 45       | 0      | 0      | 0        |
| 46       | 98089  | 98080  | 999      |
| 47       | 215355 | 215341 | 1        |
| 48       | 214683 | 214645 | 0        |
| 49       | 214870 | 214909 | 0        |
| 50       | 214906 | 214837 | 0        |
| 51       | 215887 | 215965 | 0        |
| 52       | 219043 | 218387 | 0        |
| 53       | 209458 | 209342 | 0        |
| 54       | 213390 | 213553 | 0        |
| 55       | 215100 | 215093 | 0        |
| 56       | 214460 | 214249 | 0        |
| 57       | 214883 | 215026 | 609      |
| 58       | 213354 | 213398 | 0        |
| 59       | 212175 | 211942 | 0        |

Simplest reproducer is to run dnsperf with -Q against an echo server, kill the server in the middle of test, and then restart it. dnsperf will send more queries as it tries to "catch up".

```
$ yes '. A' | dnsperf -O suppress=timeout -t 1 -Q 100000 -l 10 -s 127.0.0.1 -S1 -p 5353
DNS Performance Testing Tool
Version 2.10.0

[Status] Command line: dnsperf -O suppress=timeout -t 1 -Q 100000 -l 10 -s 127.0.0.1 -S1 -p 5353
[Status] Sending queries (to 127.0.0.1:5353)
[Status] Started at: Thu Feb  2 16:12:30 2023
[Status] Stopping after 10.000000 seconds
1675350751.204563: 99992.803613
1675350752.205596: 100005.694118
1675350753.206756: 34156.378601
1675350754.207943: 0.000000
1675350755.209119: 0.000000
1675350756.210197: 0.000000
1675350757.211227: 226568.634307
1675350758.211343: 324562.350767
1675350759.211482: 114571.074621
[Status] Testing complete (time limit)

Statistics:

  Queries sent:         999995
  Queries completed:    999595 (99.96%)
  Queries lost:         400 (0.04%)

  Response codes:       NOERROR 999595 (100.00%)
  Average packet size:  request 17, response 17
  Run time (s):         10.000020
  Queries per second:   99959.300081

  Average Latency (s):  0.000160 (min 0.000005, max 0.000604)
  Latency StdDev (s):   0.000142
```

**Comment by jelu** [comment]
@pspacek I've rework #225 here: https://github.com/jelu/dnsperf/tree/qps-wait

Can you test that branch?

I've readding the sleep between queries and made it into a long opt, please check if man-page entry look OK.

**Comment by pspacek** [comment]
Generally it works, but it has off-by-one error exactly opposite to https://github.com/DNS-OARC/dnsperf/pull/225#issuecomment-1460172401 :grin: 

```
$ yes '. NS' | dnsperf -s 127.0.0.1 -p 5353 -S1 -Q 1
DNS Performance Testing Tool
Version 2.10.0

[Status] Command line: dnsperf -s 127.0.0.1 -p 5353 -S1 -Q 1
[Status] Sending queries (to 127.0.0.1:5353)
[Status] Started at: Mon Mar 27 17:07:44 2023
[Status] Stopping after 1 run through file
1679929665.262608: 1.997545
1679929666.264238: 0.998373
1679929667.265443: 0.998796
1679929668.266623: 0.998821
1679929669.267815: 0.998809
1679929670.269016: 0.998800
```

Conceptually I think magic parameter to tune `nanosleep()` call is ugly. I think better approach would be to call equivalent of `nanosleep(now)` (as in, don't sleep at all) in a loop couple times to automagically determine the threshold (at least as default, if you insist on having a parameter).

But honestly I fail to see why it is a problem, dnsperf is full of busy-wait loops :shrug:

**Comment by jelu** [comment]
> Generally it works, but it has off-by-one error exactly opposite to https://github.com/DNS-OARC/dnsperf/pull/225#issuecomment-1460172401

You can't base that on the `-Q` output, where is the full run statistics?

> But honestly I fail to see why it is a problem, dnsperf is full of busy-wait loops shrug

Your use-cases are not the normal use-cases :smile: and I added `-O qps-threshold-wait=0` for you to turn it off.

**Comment by jelu** [comment]
Sorry, I meant you can't base it on `-S`.

`-S` run in a different thread, using different timeout. It can capture two sends because `-S 1` and `-Q 1` are not synced.

**Comment by pspacek** [comment]
Ah, okay! Now I can see what you mean - and `tcpdump` confirms your suspicion.

So the my only qualm is the magic parameter vs. auto-tuning. Should I give it a try?

**Comment by jelu** [comment]
What exactly is your issue with it?

You can turn it off with `-O` and it will be disabled if `-Q` is high enough. That to me is better then using 100% CPU when testing low QPS which your change does.

busy wait != consume all CPU

**Comment by pspacek** [comment]
For me personally it's not a problem. I was just trying to find way to avoid fiddling with magic parameter as it would improve the UI for everyone.

I'm suggesting to measure how much time `nanosleep(~ 0)` actually takes as part of program init, and then disable it (i.e. auto-tune the parameter to 0) if -Q produces time interval between packets smaller than observed `nanosleep(~0)` call duration.

**Comment by jelu** [comment]
I really fail to see how measuring that would be better and less magical then to have an option you could set.

The sleep itself is probably not what's taking the most CPU in this case, it's everything else around the `perf_net_sendto()`.

But sure, let's see code.

**Comment by pspacek** [comment]
If I remember correctly `nanosleep()` is expensive like hell. Lemme sketch something for you ...

**Comment by jelu** [comment]
Same exact ns for `clock_gettime()` no matter which clock you check? That's sounds fishy or optimized away or something...

**Comment by pspacek** [comment]
I agree, have a look: 
[nanosleep_func.c.txt](https://github.com/DNS-OARC/dnsperf/files/11149922/nanosleep_func.c.txt)

I've compiled with `gcc -O2`.

**Comment by jelu** [comment]
You should probably do `-O0` when comparing.

**Comment by pspacek** [comment]
Because anyone is going to compile dnsperf with `-O0`? :-)

Just to be sure I retried with `-O0` and it does not change anything. All the `_gettime()` functions are giving the same numbers on my systems, even when compiled with `-O0`.

**Comment by pspacek** [comment]
Interestingly, this little benchmark shows that 1000 iterations of nanosleep() seems to be good enough to get the same estimate as 1M iterations. Based on that I think it would be feasible to run the nanosleep() loop automatically when -Q is specified to auto-tune the `config->qps_threshold_wait` value. It would cause << 0.1 s delay, which I don't think is significant.

**Comment by jelu** [comment]
@pspacek I smashed your test code together, how about https://github.com/jelu/dnsperf/commit/6e5ac018d35ce1f749b1159c572ac1761803b6fd ?

**Comment by pspacek** [comment]
I went to re-read and re-test https://github.com/jelu/dnsperf/tree/qps-wait and it is not working that well. It can still create very high spikes, depending on when previous timeout occurred - relative to the current second.

This is an design issue - it averages QPS over last second. Lemme explain on an example:

Test run with:
```
$ src/dnsperf -t 0.01 -p 5300 -d /tmp/qlist -l 50 -S1 -Q 10000 -O suppress=timeout
```

PCAP captured with:
```
sudo tcpdump -i lo 'dst port 5300' -w /tmp/pcap
```

Now, on localhost port 5300 we have an echo server which responds immediately. To simulate hiccup on server side, we kill the echo server and restart it couple seconds later. dnsperf timeouts queries after 0.1 sec so it keeps sending ~ 1000 QPS even when the server is down - that's okay.

Now the problem is when the server starts up again. It answers bunch of queries and unblocks all sockets, and dnsperf attempts to "catch up" with the last second - and floods the server with lots of queries at once.

Here is table from the PCAP, generated using https://gitlab.nic.cz/knot/shotgun/-/blob/master/pcap/count-packets-over-time.lua , with 0.1 sec resolution (= 1 line in the table):

Look at time 4.5 and 7.9. At 4.9 I killed the echo server, and on 7.9 I restarted it again. The second column at time 7.9 indicates that dnsperf tried hard to catch up and flooded the server with 8201 packets within 0.1 secs:

| time_s | period_packets | period_pps | total_pps |
|--------|----------------|------------|-----------|
| 0.100  | 1001           | 10010      | 10010     |
| 0.200  | 1000           | 10000      | 10005     |
| 0.300  | 1000           | 10000      | 10003     |
| 0.400  | 1000           | 10000      | 10002     |
| 0.500  | 1000           | 10000      | 10002     |
| 0.600  | 1000           | 10000      | 10001     |
| 0.700  | 1000           | 10000      | 10001     |
| 0.800  | 1000           | 10000      | 10001     |
| 0.900  | 1000           | 10000      | 10001     |
| 1.000  | 1000           | 10000      | 10001     |
| 1.100  | 1000           | 10000      | 10000     |
| 1.200  | 1000           | 10000      | 10000     |
| 1.300  | 1000           | 10000      | 10000     |
| 1.400  | 1000           | 10000      | 10000     |
| 1.500  | 1000           | 10000      | 10000     |
| 1.600  | 1000           | 10000      | 10000     |
| 1.700  | 1000           | 10000      | 10000     |
| 1.800  | 1000           | 10000      | 10000     |
| 1.900  | 1000           | 10000      | 10000     |
| 2.000  | 1000           | 10000      | 10000     |
| 2.100  | 1000           | 10000      | 10000     |
| 2.200  | 1000           | 10000      | 10000     |
| 2.300  | 1000           | 10000      | 10000     |
| 2.400  | 1000           | 10000      | 10000     |
| 2.500  | 1000           | 10000      | 10000     |
| 2.600  | 1000           | 10000      | 10000     |
| 2.700  | 1000           | 10000      | 10000     |
| 2.800  | 1000           | 10000      | 10000     |
| 2.900  | 1000           | 10000      | 10000     |
| 3.000  | 1000           | 10000      | 10000     |
| 3.100  | 1000           | 10000      | 10000     |
| 3.200  | 1000           | 10000      | 10000     |
| 3.300  | 1000           | 10000      | 10000     |
| 3.400  | 1000           | 10000      | 10000     |
| 3.500  | 1000           | 10000      | 10000     |
| 3.600  | 1000           | 10000      | 10000     |
| 3.700  | 1000           | 10000      | 10000     |
| 3.800  | 1000           | 10000      | 10000     |
| 3.900  | 1000           | 10000      | 10000     |
| 4.000  | 1000           | 10000      | 10000     |
| 4.100  | 1000           | 10000      | 10000     |
| 4.200  | 1000           | 10000      | 10000     |
| 4.300  | 1000           | 10000      | 10000     |
| 4.400  | 1000           | 10000      | 10000     |
| 4.500  | 176            | 1760       | 9817      |
| 4.600  | 100            | 1000       | 9625      |
| 4.700  | 100            | 1000       | 9441      |
| 4.800  | 100            | 1000       | 9266      |
| 4.900  | 100            | 1000       | 9097      |
| 5.000  | 100            | 1000       | 8935      |
| 5.100  | 100            | 1000       | 8779      |
| 5.200  | 100            | 1000       | 8630      |
| 5.300  | 100            | 1000       | 8486      |
| 5.400  | 100            | 1000       | 8347      |
| 5.500  | 100            | 1000       | 8214      |
| 5.600  | 100            | 1000       | 8085      |
| 5.700  | 100            | 1000       | 7960      |
| 5.800  | 100            | 1000       | 7840      |
| 5.900  | 100            | 1000       | 7724      |
| 6.000  | 100            | 1000       | 7612      |
| 6.100  | 100            | 1000       | 7504      |
| 6.200  | 100            | 1000       | 7399      |
| 6.300  | 100            | 1000       | 7297      |
| 6.400  | 100            | 1000       | 7199      |
| 6.500  | 100            | 1000       | 7104      |
| 6.600  | 100            | 1000       | 7011      |
| 6.700  | 100            | 1000       | 6921      |
| 6.800  | 100            | 1000       | 6834      |
| 6.900  | 100            | 1000       | 6750      |
| 7.000  | 100            | 1000       | 6668      |
| 7.100  | 100            | 1000       | 6588      |
| 7.200  | 100            | 1000       | 6510      |
| 7.300  | 100            | 1000       | 6435      |
| 7.400  | 100            | 1000       | 6361      |
| 7.500  | 100            | 1000       | 6290      |
| 7.600  | 100            | 1000       | 6220      |
| 7.700  | 100            | 1000       | 6152      |
| 7.800  | 100            | 1000       | 6086      |
| 7.900  | 8201           | 82010      | 7047      |
| 8.000  | 1000           | 10000      | 7084      |
| 8.100  | 1000           | 10000      | 7120      |
| 8.200  | 1000           | 10000      | 7155      |
| 8.300  | 1000           | 10000      | 7190      |
| 8.400  | 1000           | 10000      | 7223      |
| 8.500  | 1000           | 10000      | 7256      |
| 8.600  | 1000           | 10000      | 7288      |
| 8.700  | 1000           | 10000      | 7319      |
| 8.800  | 1000           | 10000      | 7349      |
| 8.900  | 1000           | 10000      | 7379      |

In other words, https://github.com/jelu/dnsperf/tree/qps-wait shifts the original problem from "tries to catch up during the whole run" to "tries to catch up within 1 sec", but the remaining fraction of 1 sec can make the QPS limit almost infinite.

The original code in https://github.com/DNS-OARC/dnsperf/pull/225 does not have this issue, so I propose to go back to that and factor in your hack to limit CPU usage. Agreed?

**Comment by jelu** [comment]
> The original code in https://github.com/DNS-OARC/dnsperf/pull/225 does not have this issue, so I propose to go back to that and factor in your hack to limit CPU usage. Agreed?

No, not fully. Problem with `packet_step`/`last_packet` is that it just assumes a packet has been sent but there are many branches where a packet is not sent.

Going to tweak my qps-wait branch.

**Comment by pspacek** [comment]
> No, not fully. Problem with `packet_step`/`last_packet` is that it just assumes a packet has been sent but there are many branches where a packet is not sent.

Well that's intentional. -Q QPS is convenient way to specify time _expected_ spacing between packets. If the packet was not sent, then well, it cannot sent anymore because it potentially violates the expected spacing, and lead to exactly the spikiness the current qps-wait branch exhibits.

If you narrow the averaging interval from 1 s down to 0.1 s it will shorten the spike-period (period when it tries to catch up to reach specified QPS within the interval). If you take narrowing to the limit you will arrive to #225 equivalent :shrug:

**Comment by jelu** [comment]
@pspacek can you pull my branch now and do that test of yours again please?

**Comment by jelu** [comment]
sorry @pspacek, yet another update to the branch :)

**Comment by pspacek** [comment]
I've tested dec926b501fb5664061f373ae3ae4cf9b445183f and it's _almost_ working. It slightly overshots or undershots the target QPS. Tested using [same procedure as before](https://github.com/DNS-OARC/dnsperf/issues/224#issuecomment-1534955560):

For simplicity just table with QPS within 0.1 s period. Please note the 1010 and 920 at the beginning - at that point I did not even stop the echo server (yet). First server kill occured around 3.0 s mark.

| time_s | period_packets |
|--------|----------------|
| 0.100  | 1000           |
| 0.200  | 1010           |
| 0.300  | 1011           |
| 0.400  | 1010           |
| 0.500  | 1010           |
| 0.600  | 1010           |
| 0.700  | 1010           |
| 0.800  | 1010           |
| 0.900  | 1010           |
| 1.000  | 920            |
| 1.100  | 1010           |
| 1.200  | 1010           |
| 1.300  | 1010           |
| 1.400  | 1010           |
| 1.500  | 1010           |
| 1.600  | 1010           |
| 1.700  | 1010           |
| 1.800  | 1010           |
| 1.900  | 1010           |
| 2.000  | 909            |
| 2.100  | 1011           |
| 2.200  | 1010           |
| 2.300  | 1010           |
| 2.400  | 1010           |
| 2.500  | 1010           |
| 2.600  | 1010           |
| 2.700  | 1010           |
| 2.800  | 1010           |
| 2.900  | 1010           |
| 3.000  | 97             |
| 3.100  | 97             |
| 3.200  | 97             |
| 3.300  | 99             |
| 3.400  | 98             |
| 3.500  | 98             |
| 3.600  | 97             |
| 3.700  | 98             |
| 3.800  | 97             |
| 3.900  | 97             |
| 4.000  | 98             |
| 4.100  | 100            |
| 4.200  | 100            |
| 4.300  | 100            |
| 4.400  | 100            |
| 4.500  | 100            |
| 4.600  | 100            |
| 4.700  | 100            |
| 4.800  | 100            |
| 4.900  | 100            |
| 5.000  | 100            |
| 5.100  | 100            |
| 5.200  | 100            |
| 5.300  | 100            |
| 5.400  | 100            |
| 5.500  | 969            |
| 5.600  | 1010           |
| 5.700  | 1010           |
| 5.800  | 1010           |
| 5.900  | 1010           |
| 6.000  | 1010           |
| 6.100  | 1010           |
| 6.200  | 1010           |
| 6.300  | 1010           |
| 6.400  | 1010           |
| 6.500  | 1010           |
| 6.600  | 1010           |
| 6.700  | 1010           |
| 6.800  | 1010           |
| 6.900  | 1010           |
| 7.000  | 912            |
| 7.100  | 1010           |
| 7.200  | 1010           |
| 7.300  | 1010           |
| 7.400  | 1010           |
| 7.500  | 1010           |
| 7.600  | 1010           |
| 7.700  | 1010           |
| 7.800  | 1010           |
| 7.900  | 1011           |
| 8.000  | 908            |
| 8.100  | 1011           |
| 8.200  | 1010           |
| 8.300  | 1010           |
| 8.400  | 1010           |
| 8.500  | 1010           |
| 8.600  | 1010           |
| 8.700  | 1010           |
| 8.800  | 1010           |
| 8.900  | 1011           |
| 9.000  | 909            |
| 9.100  | 1010           |
| 9.200  | 1010           |
| 9.300  | 1010           |
| 9.400  | 1010           |
| 9.500  | 1011           |
| 9.600  | 1010           |
| 9.700  | 358            |
| 9.800  | 100            |
| 9.900  | 100            |
| 10.000 | 100            |
| 10.100 | 100            |
| 10.200 | 100            |
| 10.300 | 100            |
| 10.400 | 100            |
| 10.500 | 100            |
| 10.600 | 100            |
| 10.700 | 100            |
| 10.800 | 100            |
| 10.900 | 100            |
| 11.000 | 100            |
| 11.100 | 100            |
| 11.200 | 100            |
| 11.300 | 100            |
| 11.400 | 100            |
| 11.500 | 100            |
| 11.600 | 100            |
| 11.700 | 699            |
| 11.800 | 1010           |
| 11.900 | 1010           |
| 12.000 | 1010           |
| 12.100 | 1010           |
| 12.200 | 1010           |
| 12.300 | 1009           |
| 12.400 | 1010           |
| 12.500 | 1010           |
| 12.600 | 1010           |
| 12.700 | 1010           |
| 12.800 | 1010           |
| 12.900 | 1010           |
| 13.000 | 912            |
| 13.100 | 1010           |
| 13.200 | 1010           |
| 13.300 | 1010           |
| 13.400 | 1010           |
| 13.500 | 1010           |
| 13.600 | 1010           |
| 13.700 | 1010           |
| 13.800 | 1010           |
| 13.900 | 1010           |
| 14.000 | 910            |
| 14.100 | 1010           |
| 14.200 | 1010           |
| 14.300 | 1010           |
| 14.400 | 1010           |
| 14.500 | 1010           |
| 14.600 | 964            |
| 14.700 | 100            |
| 14.800 | 100            |
| 14.900 | 100            |
| 15.000 | 100            |
| 15.100 | 100            |
| 15.200 | 100            |
| 15.300 | 100            |
| 15.400 | 100            |
| 15.500 | 100            |
| 15.600 | 100            |
| 15.700 | 100            |
| 15.800 | 100            |
| 15.900 | 100            |
| 16.000 | 100            |
| 16.100 | 100            |
| 16.200 | 100            |
| 16.300 | 100            |
| 16.400 | 100            |
| 16.500 | 108            |
| 16.600 | 1010           |
| 16.700 | 1010           |
| 16.800 | 1010           |
| 16.900 | 1010           |
| 17.000 | 1010           |
| 17.100 | 1011           |
| 17.200 | 1010           |
| 17.300 | 1010           |
| 17.400 | 1010           |
| 17.500 | 1010           |
| 17.600 | 1010           |
| 17.700 | 1010           |
| 17.800 | 1010           |
| 17.900 | 1010           |
| 18.000 | 911            |
| 18.100 | 1010           |
| 18.200 | 1010           |
| 18.300 | 1010           |
| 18.400 | 1010           |
| 18.500 | 1010           |
| 18.600 | 1010           |
| 18.700 | 1010           |
| 18.800 | 1010           |
| 18.900 | 1011           |
| 19.000 | 909            |
| 19.100 | 1010           |
| 19.200 | 640            |
| 19.300 | 100            |
| 19.400 | 100            |
| 19.500 | 100            |
| 19.600 | 100            |
| 19.700 | 100            |
| 19.800 | 100            |
| 19.900 | 100            |
| 20.000 | 100            |
| 20.100 | 100            |
| 20.200 | 100            |

Also, I have to say I have kind of hard time following the logic. Comments would help :-)

**Comment by jelu** [comment]
Can you list all server kill and start times please, looks to be multiple.

**Comment by pspacek** [comment]
I've pushed https://github.com/DNS-OARC/dnsperf/pull/225/commits/ad7889ba64258c1de0a1c5ecb7fa75e84c1d5b36 which IMHO is much simpler and produces less artifacts in packet rate. Results from the same test follow (just beware, echo server stop happened at different timestamps):

| time_s | period_packets |
|--------|----------------|
| 0.100  | 979            |
| 0.200  | 1000           |
| 0.300  | 1000           |
| 0.400  | 999            |
| 0.500  | 1000           |
| 0.600  | 1000           |
| 0.700  | 1000           |
| 0.800  | 1000           |
| 0.900  | 1000           |
| 1.000  | 1000           |
| 1.100  | 1000           |
| 1.200  | 1000           |
| 1.300  | 1000           |
| 1.400  | 1000           |
| 1.500  | 1000           |
| 1.600  | 1000           |
| 1.700  | 1000           |
| 1.800  | 1000           |
| 1.900  | 1000           |
| 2.000  | 1000           |
| 2.100  | 1000           |
| 2.200  | 1000           |
| 2.300  | 1000           |
| 2.400  | 1000           |
| 2.500  | 1000           |
| 2.600  | 1000           |
| 2.700  | 1000           |
| 2.800  | 1000           |
| 2.900  | 1000           |
| 3.000  | 1000           |
| 3.100  | 1000           |
| 3.200  | 1000           |
| 3.300  | 1000           |
| 3.400  | 1000           |
| 3.500  | 1000           |
| 3.600  | 1000           |
| 3.700  | 1000           |
| 3.800  | 1000           |
| 3.900  | 1000           |
| 4.000  | 1000           |
| 4.100  | 1000           |
| 4.200  | 1000           |
| 4.300  | 1000           |
| 4.400  | 1000           |
| 4.500  | 1000           |
| 4.600  | 269            |
| 4.700  | 100            |
| 4.800  | 100            |
| 4.900  | 100            |
| 5.000  | 100            |
| 5.100  | 100            |
| 5.200  | 100            |
| 5.300  | 100            |
| 5.400  | 100            |
| 5.500  | 100            |
| 5.600  | 100            |
| 5.700  | 100            |
| 5.800  | 100            |
| 5.900  | 100            |
| 6.000  | 100            |
| 6.100  | 100            |
| 6.200  | 100            |
| 6.300  | 100            |
| 6.400  | 100            |
| 6.500  | 100            |
| 6.600  | 100            |
| 6.700  | 776            |
| 6.800  | 1000           |
| 6.900  | 1000           |
| 7.000  | 1000           |
| 7.100  | 1000           |
| 7.200  | 1000           |
| 7.300  | 1000           |
| 7.400  | 1000           |
| 7.500  | 1000           |
| 7.600  | 1000           |
| 7.700  | 1000           |
| 7.800  | 1000           |
| 7.900  | 1000           |
| 8.000  | 1000           |
| 8.100  | 1000           |
| 8.200  | 1000           |
| 8.300  | 999            |
| 8.400  | 1000           |
| 8.500  | 1000           |
| 8.600  | 1000           |
| 8.700  | 1000           |
| 8.800  | 1000           |
| 8.900  | 1000           |
| 9.000  | 1000           |
| 9.100  | 1000           |
| 9.200  | 1000           |
| 9.300  | 1000           |
| 9.400  | 1000           |
| 9.500  | 1000           |
| 9.600  | 1000           |
| 9.700  | 1000           |
| 9.800  | 1000           |
| 9.900  | 1000           |
| 10.000 | 1000           |
| 10.100 | 1000           |
| 10.200 | 1000           |
| 10.300 | 999            |
| 10.400 | 749            |
| 10.500 | 100            |
| 10.600 | 100            |
| 10.700 | 100            |
| 10.800 | 100            |
| 10.900 | 100            |
| 11.000 | 100            |
| 11.100 | 100            |
| 11.200 | 100            |
| 11.300 | 100            |
| 11.400 | 100            |
| 11.500 | 100            |
| 11.600 | 100            |
| 11.700 | 100            |
| 11.800 | 100            |
| 11.900 | 100            |
| 12.000 | 100            |
| 12.100 | 100            |
| 12.200 | 100            |
| 12.300 | 100            |
| 12.400 | 303            |
| 12.500 | 1000           |
| 12.600 | 1000           |
| 12.700 | 1000           |
| 12.800 | 1000           |
| 12.900 | 1000           |
| 13.000 | 1000           |
| 13.100 | 1000           |
| 13.200 | 1000           |
| 13.300 | 999            |
| 13.400 | 1000           |
| 13.500 | 1000           |
| 13.600 | 1000           |
| 13.700 | 1000           |
| 13.800 | 1000           |
| 13.900 | 1000           |
| 14.000 | 1000           |
| 14.100 | 1000           |
| 14.200 | 1000           |
| 14.300 | 1000           |
| 14.400 | 1000           |
| 14.500 | 1000           |
| 14.600 | 1000           |
| 14.700 | 1000           |
| 14.800 | 1000           |
| 14.900 | 1000           |
| 15.000 | 1000           |
| 15.100 | 1000           |
| 15.200 | 1000           |
| 15.300 | 1000           |
| 15.400 | 1000           |
| 15.500 | 1000           |
| 15.600 | 1000           |
| 15.700 | 1000           |
| 15.800 | 1000           |
| 15.900 | 1000           |
| 16.000 | 1000           |
| 16.100 | 1000           |
| 16.200 | 1000           |
| 16.300 | 790            |
| 16.400 | 100            |
| 16.500 | 100            |
| 16.600 | 100            |
| 16.700 | 100            |
| 16.800 | 100            |
| 16.900 | 100            |
| 17.000 | 100            |
| 17.100 | 100            |
| 17.200 | 100            |
| 17.300 | 100            |
| 17.400 | 100            |
| 17.500 | 100            |
| 17.600 | 100            |
| 17.700 | 100            |
| 17.800 | 100            |
| 17.900 | 100            |
| 18.000 | 100            |
| 18.100 | 100            |
| 18.200 | 100            |
| 18.300 | 100            |
| 18.384 | 100            |

**Comment by pspacek** [comment]
> Can you list all server kill and start times please, looks to be multiple.

I kill & restart the server manually, so just look for sharp ~ 1000 -> ~ 100 drop.

**Comment by pspacek** [comment]
BTW I'm using https://github.com/isc-projects/dns-echo-user for the echo server, if you want to try it locally. [Test procedure above](https://gitlab.nic.cz/knot/shotgun/-/blob/master/pcap/count-packets-over-time.lua) has links for PCAP capture & post-processing.
